packages/fxa-customs-server/lib/dataflow.js

```diff
@@ -38,203 +40,139 @@ module.exports = function(config, log, fetchRecords) {
-    rl_email_recipient: isEmailBlockUnexpected,
-    rl_statuscheck: isIpBlockUnexpected,
-    // accountid in this instance is a uid and we don't have a UidIpBlock, trying for just a UID block
-    rl_verifycode_sourceaddress_accountid: isUidBlockUnexpected,
-  };
-
-  // Returned for testing
-  return {
-    checkForUnexpectedBlock,
-    messageHandler,
-    parseMessage,
-    processMessage,
-  };
-
   /**
-   * Ack, log, and process a message
+   * Handle a message.
    *
    * @param {Object} message
-   * @param {Function} [_processMessage=processMessage] message processing method
    */
-  function messageHandler(message, _processMessage = processMessage) {
-    message.ack();
-
-    log.info({
-      op: 'dataflow.message',
-      count: messageCount,
-      id: message.id,
-      // message.data is a Buffer, convert to a string
-      data: message.data.toString(),
-      attributes: message.attributes,
-    });
+  async function handleMessage(message) {
+    try {
+      const action = parseMessage(message);
+      const type = TYPES.get(action.indicator_type);
 
-    messageCount++;
+      let level, op;
 
-    _processMessage(message);
-  }
+      if (reportOnly || action.suggested_action === 'report') {
+        level = 'info';
+        op = 'dataflow.message.report';
+      } else if (isFresh(action.timestamp)) {
+        const records = await fetchRecords({
+          [type]: action.indicator,
+        });
 
-  /**
-   * Parse a message and check the metadata for any unexpected rate limits
-   *
-   * @param {Object} message
-   * @param {Function} [_parseMessage=parseMessage]
-   * @param {Function} [_checkForUnexpectedBlock=checkForUnexpectedBlock]
-   */
-  function processMessage(
-    message,
-    _parseMessage = parseMessage,
-    _checkForUnexpectedBlock = checkForUnexpectedBlock
-  ) {
-    let block;
+        records[type][action.suggested_action]();
 
-    try {
-      block = _parseMessage(message);
-    } catch (err) {
+        await setRecords(records);
+
+        level = 'info';
+        op = 'dataflow.message.success';
+      } else {
+        level = 'warn';
+        op = 'dataflow.message.ignore';
+      }
+
+      message.ack();
+
+      log[level]({
+        op,
+        id: message.id,
+        timestamp: action.timestamp,
+        [type]: action.indicator,
+        severity: action.severity,
+        confidence: action.confidence,
+        heuristic: action.heuristic,
+        heuristic_description: action.heuristic_description,
+        reason: action.reason,
+        suggested_action: action.suggested_action,
+        reportOnly,
+      });
+    } catch (error) {
       log.error({
-        op: 'dataflow.message.invalid',
-        reason: err.message,
+        op: 'dataflow.message.error',
+        id: message.id,
+        data: message.data,
+        error: error.message,
       });
-    }
 
-    if (block) {
-      _checkForUnexpectedBlock(block);
+      message.nack();
```

Comment:
Yep, that's my expectation.

It's the same approach we take [in the Amplitude pipeline](https://github.com/mozilla/fxa-amplitude-send/blob/a328aed55534afbd32a545c4a449657c5f99a3ac/pubsub/index.js#L138) fwiw, to push failed messages back onto the queue in case the failures are transient. Note that above there is also a condition to `ack` old messages with no action, so things won't get pushed back on to the queue indefinitely.